src/php/types/array.rs

```diff
@@ -277,202 +198,299 @@ impl<'a> Debug for ZendHashTable<'a> {
+            end: NonNull::new(unsafe { ht.arData.offset(ht.nNumUsed as isize) }),
+        }
     }
 }
 
-impl<'a> Drop for ZendHashTable<'a> {
-    fn drop(&mut self) {
-        if self.free {
-            unsafe { zend_array_destroy(self.ptr) };
+impl<'a> Iterator for Iter<'a> {
+    type Item = (u64, Option<String>, &'a Zval);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let pos = self.pos?;
+
+        if pos == self.end? {
+            return None;
         }
+
+        let bucket = unsafe { pos.as_ref() };
+        let key = unsafe { ZendString::from_ptr(bucket.key, false) }
+            .and_then(|s| s.try_into())
+            .ok();
+
+        self.pos = NonNull::new(unsafe { pos.as_ptr().offset(1) });
+
+        Some((bucket.h, key, &bucket.val))
+    }
+
+    fn count(self) -> usize
+    where
+        Self: Sized,
+    {
+        self.ht.nNumUsed as usize
     }
 }
 
-impl<'a> Clone for ZendHashTable<'a> {
-    fn clone(&self) -> Self {
-        // SAFETY: If this fails then `emalloc` failed - we are doomed anyway?
-        // `from_ptr()` checks if the ptr is null.
+/// A container used to 'own' a Zend hashtable. Dereferences to a reference to [`HashTable`].
+///
+/// When this struct is dropped, it will also destroy the internal hashtable, unless the `into_raw`
+/// function is used.
+pub struct OwnedHashTable {
+    ptr: NonNull<HashTable>,
+    drop: bool,
+}
+
+impl OwnedHashTable {
+    /// Creates a new, empty, PHP associative array.
+    pub fn new() -> Self {
+        Self::with_capacity(HT_MIN_SIZE)
+    }
+
+    /// Creates a new, empty, PHP associative array with an initial size.
+    ///
+    /// # Parameters
+    ///
+    /// * `size` - The size to initialize the array with.
+    pub fn with_capacity(size: u32) -> Self {
+        // SAFETY: PHP allocater handles the creation of the array.
         unsafe {
-            let ptr = zend_array_dup(self.ptr);
-            Self::from_ptr(ptr, true).expect("ZendHashTable cloning failed when duplicating array.")
+            let ptr = _zend_new_array(size);
+            Self::from_ptr(ptr)
         }
     }
+
+    /// Creates an owned hashtable from a hashtable pointer, which will be freed when the
+    /// resulting Rust object is dropped.
+    ///
+    /// # Parameters
+    ///
+    /// * `ptr` - Hashtable pointer.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given pointer is null.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure that the given pointer is a valid hashtable pointer, including
+    /// non-null and properly aligned.
+    pub unsafe fn from_ptr(ptr: *mut HashTable) -> Self {
+        Self {
+            ptr: NonNull::new(ptr).expect("Invalid hashtable pointer given"),
+            drop: true,
+        }
+    }
+
+    /// Returns the inner pointer to the hashtable, without destroying the
+    pub fn into_inner(mut self) -> *mut HashTable {
+        self.drop = false;
+        self.ptr.as_ptr()
+    }
```

Comment:
That said, is `self.drop` just intended for bookkeeping to prevent us from freeing memory moved out of the type? You should be able to do the same thing with `std::mem::forget`:

```suggestion
    pub fn into_inner(mut self) -> *mut HashTable {
        let ptr = self.ptr;
        std::mem::forget(self);
        ptr
    }
```

Or use `ManuallyDrop` under the hood.

Reply:
And if you don't need drop, then `OwnedHashTable` could be defined as a newtype wrapper:
```rust
pub struct OwnedHashTable(NonNull<HashTable>)
```

Reply:
Yeah, I had changed this on the `zend-str` branch after reading the `CStr` documentation